[JENKINS-43727] show assigned labels of agents

mawinter69 opened the pull request:
See [JENKINS-43727](https://issues.jenkins.io/browse/JENKINS-43727).

If enabled this monitor will show the labels of agents on the /computer
overview page.
The maximum number of labels to show is configurable. When more labels
are defined, those will be shown when hovering over the list.

This is not using model-link as this doesn't work together with the
hovering.

<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://www.jenkins.io/download/lts/#backporting-process for more.
-->


<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Show labels assigned to agents on the agent overview page


<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

mawinter69 committed:
[JENKINS-43727] show assigned labels of agents

If enabled this monitor will show the labels of agents on the /computer
overview page.
The maximum number of labels to show is configurable. When more labels
are defined, those will be shown when hovering over the list.

This is no using model-link as this doesn't work together with the
hovering.

mawinter69 committed:
use Async descriptor

fixes the test failure

mPokornyETM commented on core/src/main/resources/hudson/node_monitors/LabelsMonitor/column.jelly:
```diff
@@ -0,0 +1,39 @@
+          <j:set var="count" value="0"/>
+          <j:forEach var="entry" items="${data}">
+            <j:if test="${count &gt;= from.maxLabelCount &amp;&amp; from.maxLabelCount &gt;= 0}">
+              <a href="${rootURL}/label/${entry.name}">${entry.name}</a>&amp;#32;
```
It will be great, when you show the label description too . Ex: in toolTip text

mawinter69 replied:
done

mPokornyETM commented:
I like this idea. I have started with new plugin 'extra-node-columns-plugin' on my local development. And this was one of thinks too. I was inspired by https://github.com/jenkinsci/extra-columns-plugin
My preferred list:

+ ConfiguredExecutorsMonitor
+ getHostName (cachedHostName)
+ getDescription
+ Build Executor Status (BuildExecutorStatusMonitor)
+ countBusy
+ getEnvironment -- EnvVars cachedEnvironment
+ workspaceList
+ getLogText or getLogRecords or doProgressiveLog
+ getACL see also getAuthorizationStrategy
+ getOfflineCause
+ getConnectTime
+ real platform name (must be implemented first)
+ isLaunchSupported
+ getAssignedLabels  
+ getTiedJobs, getBuilds ?? it might to be obsolete
+ getRetentionStrategy
+ getSystemProperties
+ ListPossibleNames (IPs)

The reason for new plugin:
+ adopt all similar plugins like percentage-du-node-column-plugin or versioncolumn-plugin
+ eliminate jenkins core code and improve test run time. It is much faster to test changes in plugin rather in core.

I am not sure which way is preferred  by jenkins dev core community but when is this PR not approved in to core, we can contribute on the plugin together.

And one last think
It will be great when user can define the order of columns, because it is ordered somehow now which is not useable when you use many of columns. But this change must be done in core.

mawinter69 commented:
For the real platform there is the platformlabeler plugin that dynamically adds various labels like architecture, os,  osversion and combinations

mawinter69 commented:
> I like this idea. I have started with new plugin 'extra-node-columns-plugin' on my local development. And this was one of thinks too. I was inspired by https://github.com/jenkinsci/extra-columns-plugin My preferred list:
> 
> * ConfiguredExecutorsMonitor
> * getHostName (cachedHostName)
> * getDescription
> * Build Executor Status (BuildExecutorStatusMonitor)
> * countBusy
> * getEnvironment -- EnvVars cachedEnvironment
> * workspaceList
> * getLogText or getLogRecords or doProgressiveLog
> * getACL see also getAuthorizationStrategy
> * getOfflineCause
> * getConnectTime
> * real platform name (must be implemented first)
> * isLaunchSupported
> * getAssignedLabels
> * getTiedJobs, getBuilds ?? it might to be obsolete
> * getRetentionStrategy
> * getSystemProperties
> * ListPossibleNames (IPs)
> 
> The reason for new plugin:
> 
> * adopt all similar plugins like percentage-du-node-column-plugin or versioncolumn-plugin
> * eliminate jenkins core code and improve test run time. It is much faster to test changes in plugin rather in core.
> 

The things in your list are not Monitors in the sense that they allow to take agents offline. Also the list is so huge that it I doubt it will look good on the /computer page. For a start one could think of an extra page for this.
Might be a thing for improvement in core to have an interface NodeColumn that is implemented by NodeMonitor that is just something to display information

mawinter69 committed:
tooltip for label

add a tooltip with the description of the label

mPokornyETM commented:
I know there are not Monitors  but your changes are also not really monitor think. Thanks for feedback. I was also thinking about new page with filter options....

mPokornyETM commented on core/src/main/resources/hudson/node_monitors/LabelsMonitor/column.jelly:
```diff
@@ -0,0 +1,39 @@
+        <j:set var="count" value="0"/>
+        <j:forEach var="entry" items="${data}">
+          <j:if test="${count &lt; from.maxLabelCount || from.maxLabelCount &lt; 0}">
+              <a href="${rootURL}/label/${entry.name}" tooltip="${h.escape(entry.description)}">${entry.name}</a>&amp;#32;
```
i prefer `<br>` instead of `&amp;#32;` When you have longer labels it looks terrible. So it makes also more sense to define the max-labels-count. In that case you define 'row-height'

mawinter69 replied:
you mean to have one label per line? That will also not look so good, as it wastes the space in the rest of the table.
As it is now it will show several labels in one line and if you have more than the max the remaining are shown when you hover over the table cell.

mPokornyETM replied:
try to put this list there any you will see what is better to read. Or create a option to enable list view. So the user can define which scenario is better for him.
![image](https://user-images.githubusercontent.com/89339813/165453194-ea8bea91-fba7-4fc4-a4bd-5f69ba3c7bfb.png)

mawinter69 replied:
This is how it looks with 5 labels shown
![image](https://user-images.githubusercontent.com/17767050/165983981-349ae851-1698-4be9-80b0-0fad5d4e1c65.png)
with the labels expanded:
![image](https://user-images.githubusercontent.com/17767050/165984202-4f010f91-4e83-4fd6-8e78-0b42307fbb1c.png)

mawinter69 replied:
With maxlabels = 1 it looks like this when all labels are shown:
![image](https://user-images.githubusercontent.com/17767050/165987337-91d5391c-63c7-4ea7-8573-995f48180676.png)

mPokornyETM commented on core/src/main/resources/hudson/node_monitors/LabelsMonitor/column.jelly:
```diff
@@ -0,0 +1,39 @@
+        <j:set var="count" value="0"/>
+        <j:forEach var="entry" items="${data}">
+          <j:if test="${count &lt; from.maxLabelCount || from.maxLabelCount &lt; 0}">
+              <a href="${rootURL}/label/${entry.name}" tooltip="${h.escape(entry.description)}">${entry.name}</a>&amp;#32;
```
Use `<a class=${entry.className} model-link inside" `... to fix rendering between labels

mawinter69 committed:
fix german translation

the plural of Label is Labels in german

mawinter69 committed:
use model link, do not hover

the hovering made it impossible to use the model-link.
Now we indicate with "..." that there are additional labels and when
clicking on it we show them and all are model links.

mawinter69 committed:
reuse some styles already defined

mawinter69 committed:
fix counting

when max labels was set to 1 labels 2 and 3 were on the same line

mawinter69 committed:
fix counting

mawinter69 committed:
fix test errors and warnings

asyncmonitor
spotbugs

mawinter69 committed:
fix spotbugs issue

mawinter69 committed:
adjust help after removing the hover

mawinter69 committed:
use method to get dynamic labels

use resource for displayname and column caption

mawinter69 committed:
Merge remote-tracking branch 'lead/master' into JENKINS-43727

mawinter69 committed:
wrap in span

class of l:icon is ignored so wrap it in a span

mawinter69 committed:
Merge branch 'master' into JENKINS-43727

NotMyFault committed:
Merge branch 'master' into JENKINS-43727

alecharp commented on core/src/main/resources/hudson/node_monitors/LabelsMonitor/column.jelly:
```diff
@@ -0,0 +1,54 @@
+  <td align="right">
+    <div class="expanding-input">
+    <div class="labels">
+      <nobr>
```
would this be the reason why the column to using as much space as it can?

alecharp reviewed (commented):
I do like the feature. 

However, when the label limit is high (even the default 5), it makes the column too wide, reducing the other column too much:

![Screen Shot 2022-06-09 at 15 17 00](https://user-images.githubusercontent.com/985955/172856429-c79ef681-e84e-45e6-b404-46b8a74f32c8.png)

when the limit is low (2 for example), the column is not too wide even when expanding to see all the labels.

mawinter69 commented on core/src/main/resources/hudson/node_monitors/LabelsMonitor/column.jelly:
```diff
@@ -0,0 +1,54 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:s="/lib/form">
+  <style type="text/css">
```
This styling is inserted for each node. Where would this need to go to to avoid this?

mawinter69 commented:
> However, when the label limit is high (even the default 5), it makes the column too wide, reducing the other column too much:
> 
> when the limit is low (2 for example), the column is not too wide even when expanding to see all the labels.
I'm used to work with a big screen, there it doesn't matter. But if you think 2 is a better default that's fine for me

mawinter69 committed:
reduce default to 2

switch to hide the icon itself instead of the span wrapper now that the
class is properly passed to the icon

mawinter69 committed:
Merge branch 'JENKINS-43727' of github.com:mawinter69/jenkins into JENKINS-43727

mPokornyETM commented:
@mawinter69 great idea to clean up the PR queue.
